### PR TITLE
Add specific never search

### LINKED ISSUE #49529: It would be nice if searching for ! in the docs gave you the never type's docs
Today if you search for "!" you get a variety of results that I don't understand: https://doc.rust-lang.org/nightly/std/?search=!

It would be nice if the first result was for the "never" primitive type, which is spelled `!` (https://doc.rust-lang.org/nightly/std/primitive.never.html)

cc https://github.com/rust-lang/rust/pull/47630 https://github.com/rust-lang/rust/issues/35121

(I think this gets A-rustdoc?  I'll let someone more knowledgeable tag it...)

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #49529.

r? @QuietMisdreavus

### CODE REVIEW

**Comment by novacrazy** [comment]
Would it not be better to add some kind of generic syntax alias system?

For example, while this may fix searching for `!`, many other things like `?` and any operators (`+`, `%`, `/`, `*`, `[`/`]`, etc.) still give weird results.

**Comment by GuillaumeGomez** [comment]
This is an excellent point! I'll update this PR with these new adds.

**Comment by novacrazy** [comment]
Maybe something like this?

```javascript
query = (function() {
    switch(raw.trim()) {
        case '!': return "Never";
        case '?': return "try!";
        case '+': return "Add";
        case '-': return "Sub";
        case '[':
        case ']': return "Index";
        // and so forth
        default: return raw;
    }
})();
```

**Comment by kennytm** [comment]
`[]` can also refer to slices, and `-` could also mean the `Neg` trait. Could query support unions? 🤔

**Comment by GuillaumeGomez** [comment]
I think I'll need to introduce a new syntax. Something like 'vec|Neg' and then it'll look for the both. What do you think of it?

**Comment by kennytm** [comment]
💡 Alternative: Allow items to register symbols as an alias index, like how the primitives are presented. 

Rationale: Non-std docs don't care about `!` being the never type. Modifying the *.js infects those non-std docs.

```rust
#[doc(primitive = "never")]
#[doc(alias = "!")]
/// blah
mod prim_never { }

#[doc(primitive = "slice")]
#[doc(alias = "[]")]
mod prim_slice { }

#[doc(alias = "-")]
pub trait Neg { }

#[doc(alias = "-")]
pub trait Sub { }
```

**Comment by novacrazy** [comment]
Is there a way to make a doc attribute like that require an unstable feature flag? That’s not really my area. 

I feel like it could be abused in crates, yet it should be allowed in crates on the rare occasion users need to define their own lang items, like with `no_core` bare-metal projects.

**Comment by kennytm** [comment]
@novacrazy 

> Is there a way to make a doc attribute like that require an unstable feature flag? That’s not really my area.

Yes, that's how `#[doc(cfg/masked/spotlight)]`'s feature-gating works.

**Comment by GuillaumeGomez** [comment]
@kennytm: This is a brilliant idea. I'm just a bit afraid that people might misuse it though or that readers won't be aware of it, and therefore not use it... But this remains an excellent idea. I think I'll start implementing it.

**Comment by GuillaumeGomez** [comment]
Indeed!

**Comment by QuietMisdreavus** [comment]
+1 for adding a general mechanism. [Adding a feature gate to a doc-attribute parameter is easy](https://github.com/rust-lang/rust/pull/44026/commits/bb6de3c9ce1681ff1246c21d2a2ec537f331e258). The compiler already whitelists the doc attribute as a whole, so adding a check for a specific parameter just involves knowing the right place to put it.

> Edit: Can `no_core` crates even build documentation?

@novacrazy Yes - as long as the compiler can build it, so can rustdoc. For example, libcore is a `#![no_core]` crate, and [its docs build fine](https://doc.rust-lang.org/nightly/core/index.html).

**Comment by pietroalbini** [comment]
Blocked on #49966

**Comment by kennytm** [comment]
@pietroalbini My alternative in https://github.com/rust-lang/rust/pull/49757#issuecomment-379570974 wouldn't need #49966, though if the alternative is chosen, this PR should be closed or entirely rewritten.

**Comment by GuillaumeGomez** [comment]
@kennytm: Your doc alias can be the same for multiple types so it still strongly requires the multi-query feature.

**Comment by kennytm** [comment]
@GuillaumeGomez I don't think so, I expect there would be multiple dedicated entries named `[]` or `-` etc, and the search UI would be like:

> screenshot_2018-04-18 23 57 50_jlnaq0

**Comment by GuillaumeGomez** [comment]
Ok, here's all the implemented functionality. I finally went for last @kennytm's comment: no need for multiquery, I just add types at the beginning of the results and we're good!

cc @QuietMisdreavus

File: src/librustdoc/html/static/main.js [hunk 1]
```diff
@@ -1062,6 +1072,11 @@
                 type = matches[1].replace(/^const$/, 'constant');
                 query = query.substring(matches[0].length);
             }
+            // In case we just get a '!' as input, we can assume that the user is looking for the
+            // `Never` primitive type.
+            if (raw === '!') {
+                query = 'Never';
+            }
```
Comment (kennytm) [inline src/librustdoc/html/static/main.js]: We could delete this code now right? 😄

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js]: Oh damn! Completely forgot about it :laughing:

**Comment by GuillaumeGomez** [comment]
Ok it's ready now.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by QuietMisdreavus** [comment]
Looks good! One round of requests from me:

* Can you open up a tracking issue for the new `doc_alias` feature and reference it in your feature gate?
* Could you write up some docs in `src/doc/unstable-book/src/language-features/doc-alias.md` talking about what the attribute does?
  * Optional: add a note in `src/doc/rustdoc/src/unstable-features.md` alongside the other doc attribute stuff linking to the Unstable Book page

**Comment by QuietMisdreavus** [comment]
Great, thanks so much! r=me pending travis.

**Pull request merged**